File: 链表/92.反转链表Ⅱ.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# 使用dummy节点可以避免很多麻烦
class Solution1:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)

        cur, firstPart = head, dummy
        for i in range(left - 1):
            firstPart, cur = firstPart.next, cur.next

        # 完成逆序部分
        pre = None
        for i in range(right - left + 1):
            temp = cur.next  # temp 还有一个身份
            cur.next = pre
            pre = cur
            cur = temp
        firstPart.next.next = cur
        firstPart.next = pre

        return dummy.next

File: 链表/92.反转链表Ⅱ.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
//* 判断left和right与节点移动次数之间的关系
var reverseBetween = function (head, left, right) {
  const dummy = new ListNode(0, head);
  let groupPrev = dummy;
  let cur = head;
  for (let i = 0; i < left - 1; i++) {
    groupPrev = groupPrev.next;
    cur = cur.next;
  }
  let prev = null;
  for (let i = 0; i < right - left + 1; i++) {
    const temp = cur.next;
    cur.next = prev;
    prev = cur;
    cur = temp;
  }
  groupPrev.next.next = cur;
  groupPrev.next = prev;
  return dummy.next;
};
